### PR TITLE
Ensure LocalCertificateSelectionCallback is still called with AuthenticateAsServerAsync

### PR DESCRIPTION
**Pull request by rmkerr**
In .NET Core 2.0, we call `LocalCertificateSelectionCallback` to choose the certificate during `AuthenticateAsServerAsync`. As part of adding support for SNI, this functionality was removed. However, some users have already implemented SNI using `LocalCertificateSelectionCallback`, and we should aim to maintain compatibility for them.

This change ensures that we maintain the old behavior of `LocalCertificateSelectionCallback` when the new SNI callback is not set. When both are set, we default to the new behavior. I also added two tests to verify this behavior.

Fixes: #29110

### CODE REVIEW

**Comment by rmkerr** [comment]
@dotnet-bot test Outerloop OSX x64 Debug Build please
@dotnet-bot test Outerloop Windows x64 Debug Build please
@dotnet-bot test Outerloop Linux x64 Debug Build please

File: src/System.Net.Security/src/System/Net/Security/SslStream.cs [hunk 1]
```diff
@@ -152,17 +152,17 @@ private X509Certificate ServerCertSelectionCallbackWrapper(string targetHost)
...
         private SslAuthenticationOptions CreateAuthenticationOptions(SslServerAuthenticationOptions sslServerAuthenticationOptions)
         {
-            if (sslServerAuthenticationOptions.ServerCertificate == null && sslServerAuthenticationOptions.ServerCertificateSelectionCallback == null)
+            if (sslServerAuthenticationOptions.ServerCertificate == null && sslServerAuthenticationOptions.ServerCertificateSelectionCallback == null && _certSelectionDelegate == null)
             {
                 throw new ArgumentNullException(nameof(sslServerAuthenticationOptions.ServerCertificate));
             }
 
-            if (sslServerAuthenticationOptions.ServerCertificate != null && sslServerAuthenticationOptions.ServerCertificateSelectionCallback != null)
+            if (sslServerAuthenticationOptions.ServerCertificate != null && sslServerAuthenticationOptions.ServerCertificateSelectionCallback != null  && _certSelectionDelegate == null)
```
Comment (krwq) [inline src/System.Net.Security/src/System/Net/Security/SslStream.cs]: what if all 3 are not null?

File: src/System.Net.Security/src/System/Net/Security/SslStream.cs [hunk 2]
```diff
@@ -152,17 +152,17 @@ private X509Certificate ServerCertSelectionCallbackWrapper(string targetHost)
...
             {
                 throw new InvalidOperationException(SR.Format(SR.net_conflicting_options, nameof(ServerCertificateSelectionCallback)));
             }
 
-            var authOptions = new SslAuthenticationOptions(sslServerAuthenticationOptions);
+            var authOptions = new SslAuthenticationOptions(sslServerAuthenticationOptions, _certSelectionDelegate);
```
Comment (krwq) [inline src/System.Net.Security/src/System/Net/Security/SslStream.cs]: I'd assign it directly as we do below for ServerCertSelectionDelegate

File: src/System.Net.Security/src/System/Net/Security/SslAuthenticationOptions.cs [hunk 3]
```diff
@@ -24,15 +24,15 @@ internal SslAuthenticationOptions(SslClientAuthenticationOptions sslClientAuthen
             RemoteCertRequired = true;
             RemoteCertificateValidationCallback = sslClientAuthenticationOptions.RemoteCertificateValidationCallback;
             TargetHost = sslClientAuthenticationOptions.TargetHost;
+            CertSelectionDelegate = localCallback;
```
Comment (krwq) [inline src/System.Net.Security/src/System/Net/Security/SslAuthenticationOptions.cs]: Please remove a comment in SecureChannel.cs saying that this code never gets called (see where we call the callback)

File: src/System.Net.Security/src/System/Net/Security/SslStream.cs (hunk 1 above)
Comment (rmkerr) [reply src/System.Net.Security/src/System/Net/Security/SslStream.cs]: That's actually necessarily the case when `LocalCertificateSelectionCallback` is set -- if my understanding is correct we still need to provide a certificate to `AuthenticateAsServerAsync`, even though we will pick the actual certificate using the callback. That's actually the case tested in the first test I added.

File: src/System.Net.Security/src/System/Net/Security/SslAuthenticationOptions.cs (hunk 3 above)
Comment (rmkerr) [reply src/System.Net.Security/src/System/Net/Security/SslAuthenticationOptions.cs]: Good catch.

File: src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs [hunk 4]
```diff
@@ -51,6 +51,94 @@ public void SslStream_ClientSendsSNIServerReceives_Ok(string hostName)
...
+                options.ServerCertificate = serverCert;
+
+                var cts = new CancellationTokenSource();
+                server.AuthenticateAsServerAsync(options, cts.Token).Wait();
+
+                Assert.Equal(1, timesCallbackCalled);
+                clientJob.Wait();
+            }
+        }
+
```
Comment (krwq) [inline src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs]: Please add a test documenting what happens if callback returns null

File: src/System.Net.Security/src/System/Net/Security/SslStream.cs (hunk 1 above)
Comment (karelz) [reply src/System.Net.Security/src/System/Net/Security/SslStream.cs]: IMO we need to leave the line as it was before. This particular check should not be affected by the fact `_certSelectionDelegate` is set or not.

File: src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs [hunk 5]
```diff
@@ -51,6 +51,94 @@ public void SslStream_ClientSendsSNIServerReceives_Ok(string hostName)
...
+        [MemberData(nameof(HostNameData))]
+        public void SslStream_ServerCallbackNotSet_UsesLocalCertificateSelection(string hostName)
+        {
+            X509Certificate serverCert = Configuration.Certificates.GetSelfSignedServerCertificate();
+
+            int timesCallbackCalled = 0;
+
+            var selectionCallback = new LocalCertificateSelectionCallback((object sender, string targetHost, X509CertificateCollection localCertificates, X509Certificate remoteCertificate, string[] issuers) =>
+            {
+                timesCallbackCalled++;
```
Comment (krwq) [inline src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs]: You're not checking targetHost here - if we expect string.Empty/null please add a test for that, if we want SNI here then in SecureChannel search for call to SniHelper and pass the same info we pass to the other callback

File: src/System.Net.Security/src/System/Net/Security/SslStream.cs (hunk 2 above)
Comment (rmkerr) [reply src/System.Net.Security/src/System/Net/Security/SslStream.cs]: Assigning it directly makes things really messy, since both of these lamdas would go into the using statement. I think that this way keeps things cleaner.

Comment (krwq) [reply src/System.Net.Security/src/System/Net/Security/SslStream.cs]: (I mean to not create new arg and instead assign property/field)

Comment (rmkerr) [reply src/System.Net.Security/src/System/Net/Security/SslStream.cs]: My bad -- I thought this was at another point in the code :)

Comment (krwq) [reply src/System.Net.Security/src/System/Net/Security/SslStream.cs]: If this method wasn't constructing SslAuthenticationOptions instance I'd agree :) logic in this class is slightly convoluted so I understand the confusion 😄

File: src/System.Net.Security/src/System/Net/Security/SslStream.cs (hunk 1 above)
Comment (rmkerr) [reply src/System.Net.Security/src/System/Net/Security/SslStream.cs]: Leaving this line as it was before would mean throwing an exception when both methods are used instead of defaulting to the new one. That isn't an unreasonable behavior, but it isn't what we originally discussed.

File: src/System.Net.Security/src/System/Net/Security/SslStream.cs (hunk 2 above)
Comment (karelz) [reply src/System.Net.Security/src/System/Net/Security/SslStream.cs]: And we should set it only if `sslServerAuthenticationOptions.ServerCertificate == null && sslServerAuthenticationOptions.ServerCertificateSelectionCallback == null`, otherwise we can dump the value here (we should add comment that it is compat-only for using old-style .ctor overload which worked also on Server

**Comment by rmkerr** [comment]
I talked to @karelz about this issue, and have decided to make the two APIs exclusive. If both are used at the same time we will throw an exception rather than defaulting to the new behavior. This maintains compat, but keeps it clear that it doesn't make sense to set bot

**Comment by davidsh** [comment]
>If both are used at the same time we will throw an exception

This probably needs to be an exception thrown during the .AuthenticateAsServerAsync(SslServerAuthenticationOptions) call.  It should probably be an "InvalidOperationException" with suitable error message.

**Comment by rmkerr** [comment]
@davidsh -- I'm working on tests now, but so far that's exactly what happens.

File: src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs (hunk 5 above)
Comment (kdg3737) [reply src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs]: You're also not checking localCertificates which should contain the initial certificate.

Comment (rmkerr) [reply src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs]: I'll add both checks -- thanks for pointing this out.

File: src/System.Net.Security/src/System/Net/Security/SslAuthenticationOptions.cs (hunk 3 above)
Comment (rmkerr) [reply src/System.Net.Security/src/System/Net/Security/SslAuthenticationOptions.cs]: I've added the comment you suggested in the latest commit.

File: src/System.Net.Security/src/System/Net/Security/SslStream.cs (hunk 1 above)
Comment (rmkerr) [reply src/System.Net.Security/src/System/Net/Security/SslStream.cs]: I've removed my change in the latest commit -- this line now ensures that we throw an exception when both callbacks are set.

File: src/System.Net.Security/src/System/Net/Security/SecureChannel.cs [hunk 6]
```diff
@@ -632,6 +632,10 @@ private bool AcquireServerCredentials(ref byte[] thumbPrint, byte[] clientHello)
             X509Certificate localCertificate = null;
             bool cachedCred = false;
 
+            // There are three options for selecting the server certificate. When 
+            // selecting which to use, we prioritize the new ServerCertSelectionDelegate
+            // API. If the new API isn't used we call LocalCertSelectionCallback, and if
```
Comment (karelz) [inline src/System.Net.Security/src/System/Net/Security/SecureChannel.cs]: Add comment to LocalCertSelectionCallback that it is there only for compat with .NET Framework only.

Also we should mention that only LocalCertSelectionCallback and ServerCertificate can be set at once.

File: src/System.Net.Security/src/System/Net/Security/SslStream.cs [hunk 7]
```diff
@@ -152,7 +152,7 @@ private X509Certificate ServerCertSelectionCallbackWrapper(string targetHost)
 
         private SslAuthenticationOptions CreateAuthenticationOptions(SslServerAuthenticationOptions sslServerAuthenticationOptions)
         {
-            if (sslServerAuthenticationOptions.ServerCertificate == null && sslServerAuthenticationOptions.ServerCertificateSelectionCallback == null)
+            if (sslServerAuthenticationOptions.ServerCertificate == null && sslServerAuthenticationOptions.ServerCertificateSelectionCallback == null && _certSelectionDelegate == null)
             {
                 throw new ArgumentNullException(nameof(sslServerAuthenticationOptions.ServerCertificate));
             }
```
Comment (karelz) [inline src/System.Net.Security/src/System/Net/Security/SslStream.cs]: We're missing check for ServerCertificateSelectionCallback & _certSelectionDelegate cannot be set both at the same time.

File: src/System.Net.Security/src/System/Net/Security/SecureChannel.cs [hunk 8]
```diff
@@ -642,7 +646,6 @@ private bool AcquireServerCredentials(ref byte[] thumbPrint, byte[] clientHello)
                     throw new AuthenticationException(SR.net_ssl_io_no_server_cert);
                 }
             }
-            // This probably never gets called as this is a client options delegate
             else if (_sslAuthenticationOptions.CertSelectionDelegate != null)
             {
```
Comment (karelz) [inline src/System.Net.Security/src/System/Net/Security/SecureChannel.cs]: What is .NET Framework behavior when CertSelectionDelegate returns null? Do we fallback to ServerCertificate?
Or is ServerCertificate basically useless whenever user sets CertSelectionDelegate?

File: src/System.Net.Security/src/System/Net/Security/SslStream.cs (hunk 7 above)
Comment (rmkerr) [reply src/System.Net.Security/src/System/Net/Security/SslStream.cs]: The existing checks already cover that case, and throw:
```
System.InvalidOperationException : The 'ServerCertificateSelectionCallback' option was already set in the SslStream constructor.
```

Comment (rmkerr) [reply src/System.Net.Security/src/System/Net/Security/SslStream.cs]: Actually, it looks like that isn't right. I based this check off the fact that all of the methods for starting a connection require a cert, but with the new SslServerAuthenticationOptions class that actually isn't the case. It looks like I do need to explicitly check for this case. I'll add that.

File: src/System.Net.Security/src/System/Net/Security/SecureChannel.cs (hunk 8 above)
Comment (krwq) [reply src/System.Net.Security/src/System/Net/Security/SecureChannel.cs]: From a response under my comment it seems like we should have a test documenting that - will be reviewing new version soon

File: src/System.Net.Security/src/System/Net/Security/SslStream.cs (hunk 7 above)
Comment (krwq) [reply src/System.Net.Security/src/System/Net/Security/SslStream.cs]: I think the other check should be something like `if (newCertCallbackSet && (oldCertCallbackSet || ServerCertificatePropertySet)) error();` - previously we allowed ServerCertProp and callback now we want to still support that but error if you try to use new and old way at the same time

File: src/System.Net.Security/src/System/Net/Security/SecureChannel.cs (hunk 8 above)
Comment (krwq) [reply src/System.Net.Security/src/System/Net/Security/SecureChannel.cs]: I agree with @karelz we should define `null` behavior but if this is what we did on full framework I'm fine with that on this PR

File: src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs [hunk 9]
```diff
@@ -51,6 +51,133 @@ public void SslStream_ClientSendsSNIServerReceives_Ok(string hostName)
...
+        [MemberData(nameof(HostNameData))]
+        public void SslStream_ServerCallbackNotSet_UsesLocalCertificateSelection(string hostName)
+        {
+            X509Certificate serverCert = Configuration.Certificates.GetSelfSignedServerCertificate();
+
+            int timesCallbackCalled = 0;
+
+            var selectionCallback = new LocalCertificateSelectionCallback((object sender, string targetHost, X509CertificateCollection localCertificates, X509Certificate remoteCertificate, string[] issuers) =>
+            {
+                Assert.Equal(targetHost, string.Empty);
```
Comment (krwq) [inline src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs]: Just making sure this is expected. `targetHost` is capable of working if you just add 2 lines of codes in SecureChannel.cs

File: src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs [hunk 10]
```diff
@@ -51,6 +51,133 @@ public void SslStream_ClientSendsSNIServerReceives_Ok(string hostName)
...
+            {
+                Task clientJob = Task.Run(() => {
+                    client.AuthenticateAsClient(hostName);
+                });
+
+                SslServerAuthenticationOptions options = DefaultServerOptions();
+                options.ServerCertificate = serverCert;
+
+                var cts = new CancellationTokenSource();
+                Assert.Throws<NotSupportedException>(WithAggregateExceptionUnwrapping(() =>
```
Comment (krwq) [inline src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs]: Is `NotSupportedException` what we got on full framework? If yes we can make it a separate issue. `AuthenticationException` would make more sense though

File: src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs (hunk 9 above)
Comment (rmkerr) [reply src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs]: Yep. This fix is purely for compat.

File: src/System.Net.Security/src/System/Net/Security/SecureChannel.cs (hunk 8 above)
Comment (rmkerr) [reply src/System.Net.Security/src/System/Net/Security/SecureChannel.cs]: I don't know what the behavior is on framework, but the behavior I pin in the test is the same on Core 2.0

File: src/System.Net.Security/src/System/Net/Security/SslStream.cs (hunk 7 above)
Comment (rmkerr) [reply src/System.Net.Security/src/System/Net/Security/SslStream.cs]: Fixed in the latest commit. See the change to line 160.

File: src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs [hunk 11]
```diff
@@ -51,6 +51,132 @@ public void SslStream_ClientSendsSNIServerReceives_Ok(string hostName)
...
+            VirtualNetwork vn = new VirtualNetwork();
+            using (VirtualNetworkStream serverStream = new VirtualNetworkStream(vn, isServer: true),
+                                        clientStream = new VirtualNetworkStream(vn, isServer: false))
+            using (SslStream server = new SslStream(serverStream, false, null, selectionCallback),
+                             client = new SslStream(clientStream, leaveInnerStreamOpen: false, validationCallback))
+            {
+                Task clientJob = Task.Run(() => {
+                    client.AuthenticateAsClient(hostName);
+                    Assert.True(false, "RemoteCertificateValidationCallback called when AuthenticateAsServerAsync was expected to fail.");
+                });
```
Comment (davidsh) [inline src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs]: I'm a little worried about this test randomly failing due to the inherent race condition you are setting up here by using Task.Run() to start the client-side of the test.  You're assuming that this will run immediately/concurrently when the latter AuthenticateAsServerAsync code runs.  Do you think this will be a problem?

Comment (krwq) [reply src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs]: It's worth validating. I think the virtual connection and also regular sockets will wait for data and this will always work correctly but it makes sense to verify if that's the case (i.e. run 1000 threads doing that in a loop for 30s) - if it fails we might want to check my tests as well :)

**Comment by rmkerr** [comment]
@dotnet-bot dotnet-bot Linked test Outerloop OSX x64 Debug Build please
@dotnet-bot dotnet-bot Linked test Outerloop Windows x64 Debug Build please
@dotnet-bot dotnet-bot Linked test Outerloop Linux x64 Debug Build please
@dotnet-bot Test Outerloop UWP CoreCLR x64 Debug Build
@dotnet-bot Test Outerloop NETFX x86 Release Build

File: src/System.Net.Security/src/System/Net/Security/SecureChannel.cs [hunk 12]
```diff
@@ -632,6 +632,12 @@ private bool AcquireServerCredentials(ref byte[] thumbPrint, byte[] clientHello)
             X509Certificate localCertificate = null;
             bool cachedCred = false;
 
+            // There are three options for selecting the server certificate. When 
+            // selecting which to use, we prioritize the new ServerCertSelectionDelegate
+            // API. If the new API isn't used we call LocalCertSelectionCallback, and if 
+            // neither is set we fall back to using ServerCertificate. When choosing
+            // from multiple certificates ServerCertSelectionDelegate is the preferred
+            // API. LocalCertSelectionCallback is maintained only for compat reasons.
```
Comment (stephentoub) [inline src/System.Net.Security/src/System/Net/Security/SecureChannel.cs]: If ServerCertSelectionDelegate returns null, should we use ServerCertificate if it was provided? (I'm actually asking; I'm not sure if there's a use case for that or not.)

File: src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs [hunk 13]
```diff
@@ -51,6 +51,132 @@ public void SslStream_ClientSendsSNIServerReceives_Ok(string hostName)
...
+                SslServerAuthenticationOptions options = DefaultServerOptions();
+                options.ServerCertificateSelectionCallback = (sender, actualHostName) =>
+                {
+                    timesCallbackCalled++;
+                    Assert.Equal(hostName, actualHostName);
+                    return serverCert;
+                };
+
+                var cts = new CancellationTokenSource();
+                Assert.Throws<InvalidOperationException>(() => {server.AuthenticateAsServerAsync(options, cts.Token).Wait();});
```
Comment (stephentoub) [inline src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs]: Couple of things:
1) The CTS never has cancellation requested, so you can just pass in CancellationToken.None instead of creating a CTS and passing in its Token.
2) I'd prefer we avoid doing synchronous blocking like this.  Please change it to something like:
```C#
await Assert.ThrowsAsync<InvalidOperationException(() => server.AuthenticateAsServerAsync(options, CancellationToken.None));
```

File: src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs [hunk 14]
```diff
@@ -51,6 +51,132 @@ public void SslStream_ClientSendsSNIServerReceives_Ok(string hostName)
...
+            {
+                Task clientJob = Task.Run(() => {
+                    client.AuthenticateAsClient(hostName);
+                });
+
+                SslServerAuthenticationOptions options = DefaultServerOptions();
+                options.ServerCertificate = serverCert;
+
+                var cts = new CancellationTokenSource();
+                server.AuthenticateAsServerAsync(options, cts.Token).Wait();
```
Comment (stephentoub) [inline src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs]:
```C#
await server.AuthenticateAsServerAsync(options, CancellationToken.None);
```

File: src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs [hunk 15]
```diff
@@ -51,6 +51,132 @@ public void SslStream_ClientSendsSNIServerReceives_Ok(string hostName)
...
+
+                Assert.Equal(1, timesCallbackCalled);
+                clientJob.Wait();
```
Comment (stephentoub) [inline src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs]:
```C#
await clientJob;
```

Comment (stephentoub) [reply src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs]: And, actually, since the two awaits are related (such that if one failed the other might hang), it'd be better to do:
```C#
await TaskTimeoutExtensions.WhenAllOrAnyFailed(new[] { clientJob, server.AuthenticateAsServerAsync(options, CancellationToken.None) });
```

File: src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs [hunk 16]
```diff
@@ -51,6 +51,132 @@ public void SslStream_ClientSendsSNIServerReceives_Ok(string hostName)
...
+                Task clientJob = Task.Run(() => {
+                    client.AuthenticateAsClient(hostName);
+                });
+
+                SslServerAuthenticationOptions options = DefaultServerOptions();
+                options.ServerCertificate = serverCert;
+
+                var cts = new CancellationTokenSource();
+                Assert.Throws<NotSupportedException>(WithAggregateExceptionUnwrapping(() =>
+                    server.AuthenticateAsServerAsync(options, cts.Token).Wait()
```
Comment (stephentoub) [inline src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs]: Same .Wait vs await comments.

File: src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs (hunk 10 above)
Comment (stephentoub) [reply src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs]: > According to @kdg3737 this throws NotSupportedException on framework, so this should be compatible

Do we have a test that runs on netfx that validates that?

File: src/System.Net.Security/src/System/Net/Security/SecureChannel.cs (hunk 12 above)
Comment (davidsh) [reply src/System.Net.Security/src/System/Net/Security/SecureChannel.cs]: Is "ServerCertificate" the parameter passed to the public API [AuthenticateAsServer](https://docs.microsoft.com/en-us/dotnet/api/system.net.security.sslstream.authenticateasserver?view=netcore-2.0) method?  If so, then yes, we should use that.

Comment (kdg3737) [reply src/System.Net.Security/src/System/Net/Security/SecureChannel.cs]: Look at the two options, either you throw the NotSupportedException like the framework does and may gain simpler support for LCSCB OR you default to the cert from the constructor and regain some of the functionality you lost by removing  localCertificates from the arguments of the callback (falling back to initial cert, localCertificates[0] == ServerCertificate). But it's not enough, you still don't have the initial certificate available in the callback (a way to solve this would be a public property on SslStream that returns this initial certificate). 
So no, it's probably better to go with framework behavior here.

File: src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs (hunk 13 above)
Comment (rmkerr) [reply src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs]: Thanks for the feedback. The other tests in the file are using the same pattern, so I'll update them all.

File: src/System.Net.Security/src/System/Net/Security/SecureChannel.cs (hunk 12 above)
Comment (karelz) [reply src/System.Net.Security/src/System/Net/Security/SecureChannel.cs]: We should go primarily after .NET Framework compat. Let's throw the same exception and make sure we have a test covering the scenario.

File: src/System.Net.Security/src/System/Net/Security/SecureChannel.cs (hunk 8 above)
Comment (karelz) [reply src/System.Net.Security/src/System/Net/Security/SecureChannel.cs]: Let's make sure we have a test case for it.

File: src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs [hunk 17]
```diff
@@ -51,6 +51,132 @@ public void SslStream_ClientSendsSNIServerReceives_Ok(string hostName)
...
+                server.AuthenticateAsServerAsync(options, cts.Token).Wait();
+
+                Assert.Equal(1, timesCallbackCalled);
+                clientJob.Wait();
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(HostNameData))]
+        public void SslStream_ServerCallbackReturnsNull_Throws(string hostName)
```
Comment (krwq) [inline src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs]: You might want to consider moving tests which only use old APIs to a separate file as this one does not run on netfx.

File: src/System.Net.Security/src/System/Net/Security/SslStream.cs (hunk 7 above)
Comment (karelz) [reply src/System.Net.Security/src/System/Net/Security/SslStream.cs]: @kdg3737 it's interesting idea - it may make the code a bit clearer. Given we are nearly done with 2.1, we should keep changes to minimum at this point, so let's stick with what we have for now.

File: src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs (hunk 9 above)
Comment (karelz) [reply src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs]: We should add comment that it is empty due to compat with .NET Framework.

File: src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs (hunk 17 above)
Comment (karelz) [inline src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs]: Nit: Rename to SslStream_ServerLocalCertificateSelectionCallbackReturnsNull_Throws?

File: src/System.Net.Security/src/System/Net/Security/SecureChannel.cs (hunk 12 above)
Comment (karelz) [reply src/System.Net.Security/src/System/Net/Security/SecureChannel.cs]: And the test exists: SslStream_ServerCallbackReturnsNull_Throws

File: src/System.Net.Security/src/System/Net/Security/SecureChannel.cs [hunk 18]
```diff
@@ -642,7 +648,6 @@ private bool AcquireServerCredentials(ref byte[] thumbPrint, byte[] clientHello)
                     throw new AuthenticationException(SR.net_ssl_io_no_server_cert);
                 }
             }
-            // This probably never gets called as this is a client options delegate
             else if (_sslAuthenticationOptions.CertSelectionDelegate != null)
             {
                 X509CertificateCollection tempCollection = new X509CertificateCollection();
```
Comment (karelz) [inline src/System.Net.Security/src/System/Net/Security/SecureChannel.cs]: Add explicit comment why we pass string.Empty (compat with .NET Framework) instead of hostName

File: src/System.Net.Security/src/System/Net/Security/SecureChannel.cs (hunk 6 above)
Comment (karelz) [reply src/System.Net.Security/src/System/Net/Security/SecureChannel.cs]: Nit: I would reword it to:
```c#
// There are three options for selecting the server certificate. When 
// selecting which to use, we prioritize the new ServerCertSelectionDelegate
// API. If the new API isn't used we call LocalCertSelectionCallback (for compat 
// with .NET Framework), and if neither is set we fall back to using ServerCertificate.
```

File: src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs (hunk 10 above)
Comment (rmkerr) [reply src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs]: No, but we should. I'll go ahead and create one in a different file that does not use the new APIs.

File: src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs (hunk 9 above)
Comment (rmkerr) [reply src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs]: Added in the latest commit.

**Comment by rmkerr** [comment]
@dotnet-bot dotnet-bot Linked dotnet-bot Linked test Outerloop OSX x64 Debug Build please
@dotnet-bot dotnet-bot Linked dotnet-bot Linked test Outerloop Windows x64 Debug Build please
@dotnet-bot dotnet-bot Linked dotnet-bot Linked test Outerloop Linux x64 Debug Build please
@dotnet-bot dotnet-bot Linked Test Outerloop UWP CoreCLR x64 Debug Build
@dotnet-bot dotnet-bot Linked Test Outerloop NETFX x86 Release Build

File: src/System.Net.Security/src/System/Net/Security/SecureChannel.cs (hunk 8 above)
Comment (rmkerr) [reply src/System.Net.Security/src/System/Net/Security/SecureChannel.cs]: The new test in SslStreamToStreamTest verifies this behavior.

File: src/System.Net.Security/src/System/Net/Security/SecureChannel.cs (hunk 18 above)
Comment (rmkerr) [reply src/System.Net.Security/src/System/Net/Security/SecureChannel.cs]: Added.

File: src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs (hunk 13 above)
Comment (rmkerr) [reply src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs]: I've modified all of these tests to avoid synchronous blocking. Feel free to point out anything I missed, or that I made async unnecessarily.

File: src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs (hunk 17 above)
Comment (rmkerr) [reply src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs]: I've moved this test to a test class that runs on netfx. I also renamed it, as requested by @karelz.

File: src/System.Net.Security/tests/FunctionalTests/SslStreamStreamToStreamTest.cs [hunk 19]
```diff
@@ -59,6 +59,29 @@ public abstract class SslStreamStreamToStreamTest
             }
         }
 
+        [Fact]
+        public async Task SslStream_ServerLocalCertificateSelectionCallbackReturnsNull_Throw()
```
Comment (rmkerr) [inline src/System.Net.Security/tests/FunctionalTests/SslStreamStreamToStreamTest.cs]: This test is formatted differently from the others. I did this to keep the format matching the rest of the tests in the file.

File: src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs [hunk 20]
```diff
@@ -22,43 +22,129 @@ public void SslStream_ClientSendsSNIServerReceives_Ok(string hostName)
...
+        [MemberData(nameof(HostNameData))]
+        public async void SslStream_ServerCallbackNotSet_UsesLocalCertificateSelection(string hostName)
+        {
+            X509Certificate serverCert = Configuration.Certificates.GetSelfSignedServerCertificate();
+
+            int timesCallbackCalled = 0;
+
+            var selectionCallback = new LocalCertificateSelectionCallback((object sender, string targetHost, X509CertificateCollection localCertificates, X509Certificate remoteCertificate, string[] issuers) =>
+            {
+                Assert.Equal(targetHost, string.Empty);
```
Comment (davidsh) [inline src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs]: expected parameters are listed first in these `Assert.Equal` methods (see similar code in file).  So this should be:

```c#
Assert.Equal(string.Empty, targetHost);
```

Comment (rmkerr) [reply src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs]: Thanks. I'll check over the rest of these as well.

**Comment by krwq** [comment]
@rmkerr I do not believe your test is running on netfx, currently this test file is not compiling when targetting netfx as new APIs haven't made it there yet.

**Comment by rmkerr** [comment]
@krwq -- I moved the specific test you mentioned to SslStreamToStreamTest.cs, and have confirmed that it is running against netfx. The others rely on behavior specific to the new API and don't make sense to move.

File: src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs [hunk 21]
```diff
@@ -70,29 +156,29 @@ public void SslStream_NoSniFromClient_CallbackReturnsNull()
...
+        private static Func<Task> WithAggregateExceptionUnwrapping(Func<Task> a)
         {
-            return () => {
+            return async () => {
                 try
                 {
-                    a();
+                    await a();
                 }
                 catch (AggregateException e)
```
Comment (stephentoub) [inline src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs]: This catch block is likely dead code now.  You can leave it to avoid churning more than is necessary at this point, but whereas Task.Wait() wraps the one or more exceptions in the task into an AggregateException, await just propagates the first exception.

Comment (rmkerr) [reply src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs]: Good to know, thanks

**Comment by rmkerr** [comment]
I'm not sure what's going on with this particular test run, but I do think think it's related. Rerunning to be safe. 
@dotnet-bot Test Outerloop NETFX x86 Release Build please
